### PR TITLE
Fixes to math.h functions

### PR DESCRIPTION
**Pull request by liammachado**
This fixes most of the functions in math.h and adds regressions for them.
Current known issues:
-The lround() and lrint() regressions are failing
-The fmod() regression results in Z3 memory blowup
-The regressions for the "float" versions of functions take much longer
than the "double" versions and/or result in Z3 memory blowup

### CODE REVIEW

File: share/smack/lib/smack.c [hunk 1]
```diff
@@ -300,6 +300,15 @@ void __SMACK_dummy(int v) {
 #define BINARY_OP(attrib,type,name,body) \
   function {:attrib} name.type(i1: type, i2: type) returns (type) body
 
+#define RNE_BINARY_OP(attrib,type,name,body) \
```
Comment (zvonimir) [inline share/smack/lib/smack.c]: Where will this be used after support for rounding mode gets added?

File: share/smack/include/math/math.h [hunk 2]
```diff
@@ -8,11 +8,11 @@
 float fabsf(float x);
 float fdimf(float x, float y);
 float roundf(float x);
+long lroundf(float x);
 //The following 3 functions are incomplete pending rounding mode implementation
```
Comment (zvonimir) [inline share/smack/include/math/math.h]: Now that we have rounding mode working, shouldn't this comment be addressed as well?

File: share/smack/include/math/math.h [hunk 3]
```diff
@@ -58,30 +55,24 @@ double modf(double x, double* y);
...
 int __isnegative(double x);
-int __ispositive(double x);
 int __signbit(double x);
-int signbit(double x);
 int __fpclassify(double x);
-int fpclassify(double x);
-int __finite(double x);
+int __isfinite(double x);
 
 //long doubles
```
Comment (zvonimir) [inline share/smack/include/math/math.h]: What do we do with long doubles?

File: share/smack/lib/math.c [hunk 4]
```diff
@@ -6,195 +6,183 @@
 //Check the length of pointers
 //#if ( __WORDSIZE == 64 )
 #if defined(__LP64__) || defined(_LP64)
-#define BUILD_64   1
+#define BUILD_64 1
```
Comment (zvonimir) [inline share/smack/lib/math.c]: Where is this used? Why do we need `BUILD_64`?

**Comment by zvonimir** [comment]
Just left my review. Some regressions are failing. Why?
Also, it would be good to have failing regressions in mathc folder as well. Those are always good to have as sanity checks.

**Comment by zvonimir** [comment]
Btw, your regressions are impressive. Good job.

File: share/smack/include/math/math.h (hunk 3 above)
Comment (liammachado) [reply share/smack/include/math/math.h]: SMACK currently doesn't support the long double functions.

File: share/smack/lib/math.c (hunk 4 above)
Comment (liammachado) [reply share/smack/lib/math.c]: This is currently used in smack.c to declare the lround functions [here](https://github.com/smackers/smack/blob/develop/share/smack/lib/smack.c#L1050). I'm not sure whether or not this is even needed, but it was there before I made any modifications.

File: share/smack/lib/smack.c [hunk 5]
```diff
@@ -1047,13 +1047,16 @@ void __SMACK_decls(void) {
   DECLARE(BUILTIN_RMODE_CONVERSION, i8, bvdouble, $si2fp,(_ to_fp 11 53));
   DECLARE(BUILTIN_RMODE_CONVERSION, i1, bvdouble, $si2fp,(_ to_fp 11 53));
 
-  #if BUILD_64
+  #if defined(__LP64__) || defined(_LP64) || defined (_WIN64)
```
Comment (zvonimir) [inline share/smack/lib/smack.c]: Totally minor thing, but please still fix it. You have an extra space before `(_WIN64)`.

File: share/smack/lib/smack.c [hunk 6]
```diff
@@ -1047,13 +1047,16 @@ void __SMACK_decls(void) {
...
+
     D("function {:builtin \"(_ fp.to_sbv 64)\"} $lround.bvfloat(rmode, bvfloat) returns (bv64);");
     D("function {:builtin \"(_ fp.to_sbv 64)\"} $lround.bvdouble(rmode, bvdouble) returns (bv64);");
 
   #else
+
```
Comment (zvonimir) [inline share/smack/lib/smack.c]: Remove this empty line.

File: share/smack/lib/smack.c [hunk 7]
```diff
@@ -1047,13 +1047,16 @@ void __SMACK_decls(void) {
...
     D("function {:builtin \"(_ fp.to_sbv 32)\"} $lround.bvfloat(rmode, bvfloat) returns (bv32);");
     D("function {:builtin \"(_ fp.to_sbv 32)\"} $lround.bvdouble(rmode, bvdouble) returns (bv32);");
+
```
Comment (zvonimir) [inline share/smack/lib/smack.c]: Remove this empty line.

File: share/smack/lib/smack.c [hunk 8]
```diff
@@ -1047,13 +1047,16 @@ void __SMACK_decls(void) {
   DECLARE(BUILTIN_RMODE_CONVERSION, i8, bvdouble, $si2fp,(_ to_fp 11 53));
   DECLARE(BUILTIN_RMODE_CONVERSION, i1, bvdouble, $si2fp,(_ to_fp 11 53));
 
-  #if BUILD_64
+  #if defined(__LP64__) || defined(_LP64) || defined (_WIN64)
+
```
Comment (zvonimir) [inline share/smack/lib/smack.c]: Remove this empty line.

File: share/smack/lib/smack.c [hunk 9]
```diff
@@ -1047,13 +1047,16 @@ void __SMACK_decls(void) {
...
     D("function {:builtin \"(_ fp.to_sbv 64)\"} $lround.bvfloat(rmode, bvfloat) returns (bv64);");
     D("function {:builtin \"(_ fp.to_sbv 64)\"} $lround.bvdouble(rmode, bvdouble) returns (bv64);");
```
Comment (zvonimir) [inline share/smack/lib/smack.c]: Remove this empty line.

File: share/smack/lib/math.c [hunk 10]
```diff
@@ -343,116 +333,24 @@ int __isnan(double x) {
...
-  __SMACK_code("@ := if $iszero.bvlongdouble(@) then 1bv32 else 0bv32;", ret, x);
-  return ret;
-}
-
-int __isinfl(long double x) {
-  int ret = __VERIFIER_nondet_int();
-  __SMACK_code("@ := if $isinfinite.bvlongdouble(@) then 1bv32 else 0bv32;", ret, x);
-  return ret;
+int __isfinite(double x) {
+  return !__isinf(x) & !__isnan(x);
```
Comment (keram88) [inline share/smack/lib/math.c]: Is there a reason to use `&` instead of `&&`?

File: share/smack/lib/math.c [hunk 11]
```diff
@@ -289,28 +263,44 @@ double fmax(double x, double y) {
...
 }
 
 double copysign(double x, double y) {
-  double ret = __VERIFIER_nondet_double();
-  if (__isnegative(x)^__isnegative(y))
-    __SMACK_code("@ := $fmul.bvdouble(@, -0e1023f53e11);", ret, x);
-  else
-    ret = x;
-  return ret;
+  if (__signbit(x) - __signbit(y)) {
```
Comment (keram88) [inline share/smack/lib/math.c]: Why use `-` instead of `!=`?

File: share/smack/lib/math.c [hunk 12]
```diff
@@ -3,125 +3,128 @@
...
+    fPart = __signbitf(x) ? -0.0f : 0.0f;
+  }
+  return fPart;
 }
 
-float modff(float x, float* y) {
-  *y = floorf(x);
-  return x -*y;
+float copysignf(float x, float y) {
+  if (__signbitf(x) - __signbitf(y)) {
```
Comment (keram88) [inline share/smack/lib/math.c]: And here.

**Pull request merged**